File: backend/app/schemas/campaign.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, field_validator

from app.models.enums import CampaignActionType, CampaignStatus, Platform


class CampaignBase(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    description: str | None = None
    platform: Platform = Platform.REDDIT
    action_type: CampaignActionType = CampaignActionType.UPVOTE
    target_url: HttpUrl

    @field_validator("target_url")
    @classmethod
    def validate_reddit_url(cls, value: HttpUrl) -> HttpUrl:
        host = (value.host or "").lower()
        if host not in {"reddit.com", "www.reddit.com", "old.reddit.com"}:
            raise ValueError("target_url must be a Reddit URL")
        return value


class CampaignCreate(CampaignBase):
    account_ids: list[UUID] = Field(min_length=1)


class CampaignRead(BaseModel):
    id: UUID
    name: str
    description: str | None = None
    platform: Platform
    action_type: CampaignActionType
    target_url: str
    status: CampaignStatus
    account_ids: list[UUID]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CampaignRunResult(BaseModel):
    account: str
    opened: bool
    clicked: bool
    verified: bool
    reason: str | None = None


class CampaignRunResponse(BaseModel):
    campaign: CampaignRead
    success: bool
    results: list[CampaignRunResult]

File: frontend/src/types/campaign.ts
export type CampaignStatus = "Draft" | "Ready" | "Running" | "Completed" | "Failed";
export type CampaignActionType = "UPVOTE";
export type WorkflowActionType = "OPEN_URL" | "WAIT" | "SCROLL" | "OPEN_POST" | "BACK" | "COMMENT" | "UPVOTE";

export interface Campaign {
  id: string;
  name: string;
  description: string | null;
  platform: "reddit";
  action_type: CampaignActionType;
  target_url: string;
  comment_text: string | null;
  status: CampaignStatus;
  account_ids: string[];
  created_at: string;
  updated_at: string;
}

export interface CampaignInput {
  name: string;
  description?: string | null;
  platform: "reddit";
  action_type: CampaignActionType;
  target_url: string;
  comment_text?: string | null;
  account_ids: string[];
}

export interface CampaignUpdateInput {
  comment_text?: string | null;
}

export interface CampaignRunResult {
  id?: string;
  workflow_step_id?: string;
  action_type: WorkflowActionType;
  success: boolean;
  reason: string | null;
  detail: string | null;
  verified?: boolean | null;
}

export interface WorkflowAccountResult {
  account: string;
  steps: CampaignRunResult[];
}

export interface CampaignRunResponse {
  campaign_id: string;
  success: boolean;
  results: WorkflowAccountResult[];
}

export interface WorkflowStep {
  id: string;
  campaign_id: string;
  step_order: number;
  action_type: WorkflowActionType;
  config: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface Workflow {
  campaign_id: string;
  steps: WorkflowStep[];
}

export interface WorkflowInputStep {
  action_type: WorkflowActionType;
  config: Record<string, unknown>;
}

export interface WorkflowInput {
  steps: WorkflowInputStep[];
}
